kong/plugins/rate-limiting/policies/init.lua

```diff
@@ -175,26 +276,46 @@ return {
       end
 
       return 0
-    end
+    end,
   },
   ["redis"] = {
     increment = function(conf, limits, identifier, current_timestamp, value)
-            -- the usage function already incremented the value of redis key to avoid race condition in concurrent calls
+      local periods = timestamp.get_timestamps(current_timestamp)
 
-      -- because of that we don't need to increment here
-      return true
+      if conf.sync_rate == SYNC_RATE_REALTIME then
+        -- we already incremented the counter at usage()
+        return true
+
+      else
+        update_local_counters(conf, periods, limits, identifier, value)
+        return periodical_sync(conf, sync_to_redis)
+      end
     end,
     usage = function(conf, identifier, period, current_timestamp)
+      local periods = timestamp.get_timestamps(current_timestamp)
+      local cache_key = get_local_key(conf, identifier, period, periods[period])
+
+      -- use local cache to reduce the number of redis calls
+      -- also by pass the logic of incrementing the counter
+      if conf.sync_rate ~= SYNC_RATE_REALTIME and cur_usage[cache_key] then
+        if cur_usage_expire_at[cache_key] > ngx_time() then
+          return cur_usage[cache_key] + (cur_delta[cache_key] or 0)
+        end
+
+        cur_usage[cache_key] = 0
+        cur_usage_expire_at[cache_key] = periods[period] + EXPIRATION[period]
+        cur_delta[cache_key] = 0
+
+        return 0
```

Comment:
could you elaborate if cur_usage_expire_at[cache_key] <= ngx_time(), why do we return 0 as usage?

Reply:
if `cur_usage_expire_at[cache_key] <= ngx_time()`, that means the local usage cache has already expired, so we need to reset the counter.

Reply:
To be more precise, it's not the expiration time of the cache, but the end time of a time window.
When we see an outdated time window that means we are starting another window (locally) for the first time, and the proper value for it is 0.

Reply:
And given we are starting a new window locally, it's very likely the (true value of) total counter is very close to 0. (Answering an offline question from @fffonion)